### PR TITLE
Abidlabs/similarity

### PR DESCRIPTION
**Pull request by abidlabs**
Adds embedding plot functionality in gradio library
* Creates a "view embedding plot" for interface
* Intelligently deduces what kind of labels are provided and colors the labels appropriately
* Tracks changes as data is manipulated

### CODE REVIEW

File: build/lib/gradio/inputs.py [hunk 1]
```diff
@@ -417,6 +449,15 @@ def get_interpretation_scores(self, x, neighbors, scores):
             final_scores.append(score_set)
         return final_scores
 
+    def embed(self, x):
+        if self.type == "value":
+            return [choice in x for choice in self.choices]
```
Comment (aliabid94) [inline build/lib/gradio/inputs.py:454]: shouldn't this be [float(choice in x) for choice in self.choices]

File: build/lib/gradio/inputs.py [hunk 2]
```diff
@@ -417,6 +449,15 @@ def get_interpretation_scores(self, x, neighbors, scores):
...
+        elif self.type == "index":
+            return [index in x for index in range(len(choices))]
```
Comment (aliabid94) [inline build/lib/gradio/inputs.py:456]: Same as above

File: build/lib/gradio/inputs.py [hunk 3]
```diff
@@ -469,6 +510,14 @@ def get_interpretation_scores(self, x, neighbors, scores):
         scores.insert(self.choices.index(x), None)
         return scores
 
+    def embed(self, x):
+        if self.type == "value":
+            return [choice==x for choice in self.choices]
```
Comment (aliabid94) [inline build/lib/gradio/inputs.py:515]: Same as above?

File: build/lib/gradio/interface.py [hunk 4]
```diff
@@ -42,6 +43,7 @@ def __init__(self, fn, inputs, outputs, verbose=False, examples=None,
                  title=None, description=None, thumbnail=None, 
                  server_port=None, server_name=networking.LOCALHOST_NAME,
                  allow_screenshot=True, allow_flagging=True,
+                 embedding_fn="default",
```
Comment (aliabid94) [inline build/lib/gradio/interface.py:46]: since the kwarg for interpretation is interpretation="default", maybe embedding="default" for consistency, instead of embedding_fn?

File: build/lib/gradio/embeddings.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
+import numpy as np
+
+SMALL_CONST = 1e-10
+
+def embed_text(text):
```
Comment (aliabid94) [inline build/lib/gradio/embeddings.py:5]: why is embed_text not directly inside the Textbox component? no need to be in a separate file

File: demo/digit_classifier.py [hunk 6]
```diff
@@ -14,12 +15,15 @@ def recognize_digit(image):
     prediction = model.predict(image).tolist()[0]
     return {str(i): prediction[i] for i in range(10)}
 
+im = gradio.inputs.Image(shape=(28, 28), image_mode='L', invert_colors=False)
```
Comment (aliabid94) [inline demo/digit_classifier.py:18]: why is this not sketchpad mode anymore?

File: demo/image_classifier.py [hunk 7]
```diff
@@ -37,4 +37,4 @@ def image_classifier(im):
     ])
 
 if __name__ == "__main__":
-    iface.launch()
+    iface.launch(share=True)
```
Comment (aliabid94) [inline demo/image_classifier.py:40]: remove share=True

File: gradio/embeddings.py [hunk 8]
```diff
@@ -0,0 +1,37 @@
+import numpy as np
+from sklearn.decomposition import PCA
+
+SMALL_CONST = 1e-10
+
+def embed_text(text):
```
Comment (aliabid94) [inline gradio/embeddings.py:6]: move embed_text directly into component

File: gradio/static/js/gradio.js [hunk 9]
```diff
@@ -285,6 +290,63 @@ function gradio(config, fn, target, example_file_path) {
     }
     target.find(".examples > table > tbody").html(html);
   }
+  function getBackgroundColors(){
+    //Gets the background colors for the embedding plot
+    console.log("io", io_master)
```
Comment (aliabid94) [inline gradio/static/js/gradio.js:295]: remove

File: gradio/static/js/gradio.js [hunk 10]
```diff
@@ -285,6 +290,63 @@ function gradio(config, fn, target, example_file_path) {
...
+    // If labels aren't loaded, or it's not a label output interface:
+    if (!io_master.loaded_examples || io_master["config"]["output_interfaces"][0][0]!="label") {
+      return 'rgb(54, 162, 235)'
+    }
+    // If it is a label interface, get the labels
+    let labels = []
+    let isConfidencesPresent = false;
+    for (let i=0; i<Object.keys(io_master.loaded_examples).length; i++) {
+      console.log(io_master.loaded_examples[i])
```
Comment (aliabid94) [inline gradio/static/js/gradio.js:304]: remove

File: gradio/static/js/gradio.js [hunk 11]
```diff
@@ -285,6 +290,63 @@ function gradio(config, fn, target, example_file_path) {
     }
     target.find(".examples > table > tbody").html(html);
   }
+  function getBackgroundColors(){
```
Comment (aliabid94) [inline gradio/static/js/gradio.js:293]: move this whole function to utils.js

File: gradio/static/js/gradio.js [hunk 12]
```diff
@@ -51,10 +51,14 @@ function gradio(config, fn, target, example_file_path) {
       <button class="run_examples">Run All</button>
       <button class="load_prev">Load Previous <em>(CTRL + &larr;)</em></button>
       <button class="load_next">Load Next <em>(CTRL + &rarr;)</em></button>
-      <button class="order_similar">Order by Similarity</em></button>
+      <button class="order_similar">Order by Similarity</button>
+      <button class="view_embeddings">View Embeddings</button>
```
Comment (aliabid94) [inline gradio/static/js/gradio.js:55]: will these embedding buttons and functionality be visible even if embeddings hasn't been enabled?

File: gradio/static/js/gradio.js [hunk 13]
```diff
@@ -317,6 +380,82 @@ function gradio(config, fn, target, example_file_path) {
...
     target.find(".load_next").click(next_example);
+    target.find(".order_similar").click(function() {
+      io_master.score_similarity(function() {
+        io_master.current_page = 0
+        io_master.current_example = null;
+        load_page();  
+      })
+    });
+    target.find(".view_examples").click(function() {
+      target.find(".examples > table").removeClass("invisible");
```
Comment (aliabid94) [inline gradio/static/js/gradio.js:391]: can you attach a second class to all example related elements called "example-content", and all embedded related content called "embedding-content", and just toggle the visibility of this secondary class? instead of targeting every single element

File: build/lib/gradio/inputs.py (hunk 1 above)
Comment (abidlabs) [reply build/lib/gradio/inputs.py:454]: Not strictly necessary, since python does the implicit casting, but I'll make it explicit

File: gradio/static/js/gradio.js (hunk 12 above)
Comment (abidlabs) [reply gradio/static/js/gradio.js:55]: Yes because embeddings is enabled by default when examples are
